### PR TITLE
fix: handle exceptions in callbacks

### PR DESCRIPTION
**Pull request by tadejsv**
This adds a new `on_exception` hook to the callbacks. It is used in progress bar to properly shut it down if an exception occurs.

### CODE REVIEW

**Review by bwanglzu — APPROVED** [review]
LGTM!

**Comment by tadejsv** [comment]
@bwanglzu i will add some changes first

File: finetuner/tuner/base.py [hunk 1]
```diff
@@ -186,6 +185,20 @@ def fit(
...
+                train_data,
+                eval_data,
+                epochs,
+                batch_size,
+                num_items_per_class,
+                preprocess_fn,
+                collate_fn,
+                num_workers,
+            )
+        except BaseException as e:
```
Comment (JoanFM) [inline finetuner/tuner/base.py:199]: why not simply Exception?

Comment (tadejsv) [reply finetuner/tuner/base.py:201]: I think `BaseException` is the base class of `Exception`, and some exception directly subclass `BaseException`

**Review by bwanglzu — APPROVED** [review]
LGTM, tested in notebook, worked!

**Pull request merged**